README


This archive contains the distribution Test-MinimumVersion,
version 0.101081:

  does your code require newer perl than you think?

This software is copyright (c) 2007 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




Changes
Revision history for Test-MinimumVersion

0.101081  2013-11-30 18:10:02 America/New_York
        - add a "skip" option to all_minimum_version_ok
        - use CPAN::Meta for META.* files, which should do a better
          job at figuring out prereqs

0.101080  2010-04-18 14:07:08 America/New_York
          more "comparison to undef" bugs fixed (thanks, Tux)

0.101050  2010-04-15 15:31:56 America/New_York
          convert to dzil, fix a prereq num/str conversion stupidity
          fix bug when comparing versions to undef

0.013     2010-01-17
          declare the dependency on YAML::Tiny (thanks, jawnsy)

0.012     2010-01-16
          stop inlining YAML::Tiny
          add a very, very, very basic test

0.011     2009-06-29
          update code to behave better with subtests

0.010     2009-05-31
          attempt to not croak on empty files

0.009     2009-04-18
          display version_markers when syntax requirements are too high

0.008     2008-05-22
          change default search locations for code to test
          update embedded YAML::Tiny
          packaging updates

0.007     2007-09-17
          fix t/pod-coverage.t; numerous stupid bugs!

0.006     2007-08-24
          all_version_* methods only plan, now, if there is no plan
          added no_plan option to all_version_* methods

0.005     2007-07-03
          quash some warnings
          added all_version_from_metayml_ok

0.004     2007-04-04
          add prereq for File::Find::Rule::Perl

0.003     2007-04-03
          change how it all interfaces with the programmer

0.001     2007-04-02
          initial release


lib/Test/MinimumVersion.pm
use 5.006;
use strict;
use warnings;
package Test::MinimumVersion;
BEGIN {
  $Test::MinimumVersion::VERSION = '0.101080';
}
use base 'Exporter';
# ABSTRACT: does your code require newer perl than you think?


use File::Find::Rule;
use File::Find::Rule::Perl;
use Perl::MinimumVersion 1.20; # accuracy
use YAML::Tiny 1.40; # bug fixes
use version 0.70;

use Test::Builder;
@Test::MinimumVersion::EXPORT = qw(
  minimum_version_ok
  all_minimum_version_ok
  all_minimum_version_from_metayml_ok
);

sub import {
  my($self) = shift;
  my $pack = caller;

  my $Test = Test::Builder->new;

  $Test->exported_to($pack);
  $Test->plan(@_);

  $self->export_to_level(1, $self, @Test::MinimumVersion::EXPORT);
}

sub _objectify_version {
  my ($version) = @_;
  $version = eval { $version->isa('version') } 
           ? $version
           : version->new($version);
}


sub minimum_version_ok {
  my ($file, $version) = @_;

  my $Test = Test::Builder->new;

  $version = _objectify_version($version);

  my $pmv = Perl::MinimumVersion->new($file);

  my $explicit_minimum = $pmv->minimum_explicit_version || 0;
  my $minimum = $pmv->minimum_syntax_version($explicit_minimum) || 0;

  my $is_syntax = 1
    if $minimum and $minimum > $explicit_minimum;

  $minimum = $explicit_minimum
    if $explicit_minimum and $explicit_minimum > $minimum;

  my %min = $pmv->version_markers;

  if ($minimum <= $version) {
    $Test->ok(1, $file);
  } else {
    $Test->ok(0, $file);
    $Test->diag(
      "$file requires $minimum "
      . ($is_syntax ? 'due to syntax' : 'due to explicit requirement')
    );

    if ($is_syntax and my $markers = $min{ $minimum }) {
      $Test->diag("version markers for $minimum:");
      $Test->diag("- $_ ") for @$markers;
    }
  }
}


sub all_minimum_version_ok {
  my ($version, $arg) = @_;
  $arg ||= {};
  $arg->{paths} ||= [ qw(lib t xt/smoke), glob ("*.pm"), glob ("*.PL") ];

  my $Test = Test::Builder->new;

  $version = _objectify_version($version);

  my @perl_files;
  for my $path (@{ $arg->{paths} }) {
    if (-f $path and -s $path) {
      push @perl_files, $path;
    } elsif (-d $path) {
      push @perl_files, File::Find::Rule->perl_file->in($path);
    }
  }

  unless ($Test->has_plan or $arg->{no_plan}) {
    $Test->plan(tests => scalar @perl_files);
  }

  minimum_version_ok($_, $version) for @perl_files;
}


sub all_minimum_version_from_metayml_ok {
  my ($arg) = @_;
  $arg ||= {};

  my $Test = Test::Builder->new;

  $Test->plan(skip_all => "META.yml could not be found")
    unless -f 'META.yml' and -r _;

  my $documents = YAML::Tiny->read('META.yml');

  $Test->plan(skip_all => "no minimum perl version could be determined")
    unless my $version = $documents->[0]->{requires}{perl};

  all_minimum_version_ok($version, $arg);
}

1;

__END__
=pod

=head1 NAME

Test::MinimumVersion - does your code require newer perl than you think?

=head1 VERSION

version 0.101080

=head1 SYNOPSIS

Example F<minimum-perl.t>:

  #!perl
  use Test::MinimumVersion;
  all_minimum_version_ok('5.008');

=head1 FUNCTIONS

=head2 minimum_version_ok

  minimum_version_ok($file, $version);

This test passes if the given file does not seem to require any version of perl
newer than C<$version>, which may be given as a version string or a version
object.

=head2 all_minimum_version_ok

  all_minimum_version_ok($version, \%arg);

Given either a version string or a L<version> object, this routine produces a
test plan (if there is no plan) and tests each relevant file with
C<minimum_version_ok>.

Relevant files are found by L<File::Find::Rule::Perl>.

C<\%arg> is optional.  Valid arguments are:

  paths   - in what paths to look for files; defaults to (t, lib, xt/smoke,
            and any .pm or .PL files in the current working directory)
            if it contains files, they will be checked
  no_plan - do not plan the tests about to be run

=head2 all_minimum_version_from_metayml_ok

  all_minimum_version_from_metayml_ok(\%arg);

This routine checks F<META.yml> for an entry in F<requires> for F<perl>.  If no
META.yml file or no perl version is found, all tests are skipped.  If a version
is found, the test proceeds as if C<all_minimum_version_ok> had been called
with that version.

=head1 AUTHOR

  Ricardo Signes

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Ricardo Signes.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



inc/Module/Install/JSONMETA.pm
#line 1
use strict;
use warnings;
package Module::Install::JSONMETA;
use Module::Install::Base;

BEGIN {
  our @ISA = qw(Module::Install::Base);
  our $ISCORE  = 1;
  our $VERSION = '5.000';
}

#line 16

our $json;
sub jsonmeta {
  my ($self) = @_;

  unless (eval { require JSON; JSON->VERSION(2); 1 }) {
    die "could not load JSON.pm version 2 or better; can't use jsonmeta\n"
      if $self->is_admin;
    return; # Not admin, just silently ignore.
  }

  $self->_hook_yaml_tiny;
  $self->_hook_admin_metadata;
  
}

sub _hook_yaml_tiny { 

  require YAML::Tiny;

  $json ||= JSON->new->ascii(1)->pretty;
  
  no warnings 'redefine';

  *YAML::Tiny::Dump = sub {
    my $data = shift;
    $data->{generated_by} &&= $data->{generated_by}
                            . ' using ' . __PACKAGE__
                            .  ' version '
                            .  __PACKAGE__->VERSION;
    $json->encode($data) . "\n"
  };

  *YAML::Tiny::Load = sub {
    $json->decode(shift);
  };

  *YAML::Tiny::LoadFile = sub {
    open FILE, "<", shift or die $!;
    local $/;
    my $data = <FILE>;
    close FILE;
    $json->decode($data);
  }
}

sub _hook_admin_metadata {
  my $mi = shift;

  return unless $mi->is_admin;

  no warnings 'redefine';
  *Module::Install::Admin::Metadata::read_meta = sub {
    my $self = shift;

    # Admin time only, so this should be okay to just die
    require YAML::Tiny;

    my @docs = YAML::Tiny::LoadFile('META.json');
    return $docs[0];
  };

  *Module::Install::Admin::Metadata::meta_generated_by_us = sub {
    my ($self, $req_version) = @_;

    my $meta = $self->read_meta;
    my $want  = ref($self->_top);

    $want .= " version " . $req_version
      if defined $req_version;

    $DB::single = 1;
    return $meta->{generated_by} =~ /^\Q$want\E/;
  };

  *Module::Install::Admin::Metadata::remove_meta = sub {
    my $self = shift;
    my $ver  = $self->_top->VERSION;

    return unless -f 'META.json';
    return unless $self->meta_generated_by_us($ver);
    unless (-w 'META.json') {
      warn "Can't remove META.json file. Not writable.\n";
      return;
    }
    warn "Removing auto-generated META.json\n";
    unless ( unlink 'META.json' ) {
      die "Couldn't unlink META.json:\n$!";
    }
    return;
  };

  *Module::Install::Admin::Metadata::write_meta = sub {
    my $self = shift;
    if ( -f "META.json" ) {
      return unless $self->meta_generated_by_us();
    } else {
      $self->clean_files('META.json');
    }
    print "Writing META.json\n";
    Module::Install::_write("META.json", $self->dump_meta);
    return;
  };
}

1;


Makefile.PL

use strict;
use warnings;

use 5.008000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "does your code require newer perl than you think?",
  "AUTHOR" => "Ricardo Signes",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Test-MinimumVersion",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Test::MinimumVersion",
  "PREREQ_PM" => {
    "CPAN::Meta" => 0,
    "Exporter" => 0,
    "File::Find::Rule" => 0,
    "File::Find::Rule::Perl" => 0,
    "Perl::MinimumVersion" => "1.32",
    "Test::Builder" => 0,
    "base" => 0,
    "strict" => 0,
    "version" => "0.70",
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => "0.96",
    "Test::Tester" => 0
  },
  "VERSION" => "0.101081",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CPAN::Meta" => 0,
  "Exporter" => 0,
  "File::Find::Rule" => 0,
  "File::Find::Rule::Perl" => 0,
  "Perl::MinimumVersion" => "1.32",
  "Test::Builder" => 0,
  "Test::More" => "0.96",
  "Test::Tester" => 0,
  "base" => 0,
  "strict" => 0,
  "version" => "0.70",
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



